src/options/parser.rs

```diff
@@ -1,815 +1,248 @@
-/// A **long argument** is a string. This can be a UTF-8 string, even though
-/// the arguments will all be unchecked `OsString` values, because we don’t
-/// actually store the user’s input after it’s been matched to a flag, we just
-/// store which flag it was.
-pub type LongArg = &'static str;
-
-/// A **list of values** that an option can have, to be displayed when the
-/// user enters an invalid one or skips it.
-///
-/// This is literally just help text, and won’t be used to validate a value to
-/// see if it’s correct.
-pub type Values = &'static [&'static str];
-
-/// A **flag** is either of the two argument types, because they have to
-/// be in the same array together.
-#[derive(PartialEq, Eq, Debug, Copy, Clone)]
-pub enum Flag {
-    Short(ShortArg),
-    Long(LongArg),
-}
-
-impl Flag {
-    pub fn matches(&self, arg: &Arg) -> bool {
-        match self {
-            Self::Short(short) => arg.short == Some(*short),
-            Self::Long(long) => arg.long == *long,
-        }
-    }
-}
-
-impl fmt::Display for Flag {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
-        match self {
-            Self::Short(short) => write!(f, "-{}", *short as char),
-            Self::Long(long) => write!(f, "--{long}"),
+pub use clap::Parser;
+use std::ffi::OsString;
+
+#[derive(Parser, Default)]
+#[command(author, version, about, long_about)] // Read from `Cargo.toml`
+#[clap(disable_help_flag = true)]
+pub struct Opts {
+    pub paths: Vec<OsString>,
+    /// Show hidden files.
+    #[arg(short, long, action = clap::ArgAction::Count)]
+    pub all: u8,
+    /// display extended file metadata as a table.
+    #[arg(short, long, action = clap::ArgAction::Count)]
+    pub long: u8,
+    /// list each file's Git status, if tracked or ignored.
+    #[arg(long, action = clap::ArgAction::Count)]
+    pub git: u8,
+    /// Display one entry per line.
+    #[arg(short = '1', long, action = clap::ArgAction::Count)]
+    pub oneline: u8,
+    ///recurse into directories as a tree.
+    #[arg(short = 'T', long, action = clap::ArgAction::Count)]
+    pub tree: u8,
+    /// display entries as a grid (default).
+    #[arg(short = 'G', long, action = clap::ArgAction::Count)]
+    pub grid: u8,
+    /// sort the grid across, rather than downwards.
+    #[arg(short = 'x', long, action = clap::ArgAction::Count)]
+    pub across: u8,
+    /// recurse into directories.
+    #[arg(short = 'R', long, action = clap::ArgAction::Count)]
+    pub recurse: u8,
+    /// display type indicator by file names.
+    #[arg(short = 'F', long, action = clap::ArgAction::Count)]
+    pub classify: u8,
+    #[arg(short = 'X', long, action = clap::ArgAction::Count)]
+    pub dereference: u8,
+    /// set screen width in columns.
+    #[arg(short = 'w', long)]
+    pub width: Option<usize>,
+    /// when to use terminal colours (always, auto, never).
+    #[arg(long)]
+    pub color: Option<OsString>,
+    #[arg(long)]
+    pub colour: Option<OsString>,
```

Comment:
So it seems we might be able to use Enums here instead of a bunch of strings 

https://docs.rs/clap/latest/clap/_derive/_tutorial/chapter_3/index.html#enumerated-values

Reply:
Didn't realize Terts had already suggested that and provided an example :+1: